### PR TITLE
Dashboard: Popover Panel UI Component

### PR DESCRIPTION
**Pull request by BrittanyIRL**
# Overview
- Adding functionality of popover panel dropdown type (used on templates as checkbox filtering). 
- Fixing pills to be tab-able for a11y (using same approach as story editor did with their toggle) 

## Updates 
- the popover panel is a type of dropdown. It functions slightly different though. Added a check on menuItemSelect for panel type dropdown, if panel then no auto close so can select multiple options. Only way to close is the 'x' or clicking out of the panel. 
- Since items was already required for other types of dropdowns, adjusted the format for the panel to follow suit - only difference is that labels can be nodes since we might need an icon or something with the label text.

popoverdemo

## Outstanding questions
- not sure what the max width of this panel should be, figma's example shows 595px so I went with that.

## Todo
- [x]  consolidate item propTyps

### CODE REVIEW

File: assets/src/dashboard/components/dropdown/stories/index.js [hunk 1]
```diff
@@ -93,24 +93,47 @@ export const _transparent = () => {
...
 export const _panel = () => {
+  const [statefulDemoData, setStatefulDemoData] = useState(categoryDemoData);
+
+  const updateDemoDataState = (datumToUpdate) => {
+    const newDemoData = statefulDemoData.map((item) => {
+      if (item.value === datumToUpdate) {
+        return { ...item, selected: !item.selected };
+      }
+      return item;
+    });
```
Comment (BrittanyIRL) [inline assets/src/dashboard/components/dropdown/stories/index.js]: this isn't perfect, just demoing the hook up of `onChange`

File: assets/src/dashboard/components/dropdown/stories/index.js [hunk 2]
```diff
@@ -93,24 +93,47 @@ export const _transparent = () => {
...
+    value: 'all',
+    selected: true,
+  },
+  { label: 'Arts and Crafts', value: 'arts_crafts' },
+  { label: 'Beauty', value: 'beauty' },
+  { label: 'Cooking', value: 'cooking' },
+  { label: 'News', value: 'news' },
+  { label: 'Sports', value: 'sports' },
+  { label: 'News', value: 'news' },
+  { label: 'UNCLICKABLE', value: false },
```
Comment (carloskelly13) [inline assets/src/dashboard/components/dropdown/stories/index.js]: For these we can still use the `__()` Wordpress i18n helper

File: assets/src/dashboard/components/dropdown/stories/index.js (hunk 1 above)
Comment (carloskelly13) [reply assets/src/dashboard/components/dropdown/stories/index.js]: We can wrap this in a `useCallback`

**Review by carloskelly13 — CHANGES_REQUESTED** [review]
This looks good, just a couple functional changes.

File: assets/src/dashboard/components/dropdown/stories/index.js (hunk 2 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/components/dropdown/stories/index.js]: even though it's just dummy content in storybook?

File: assets/src/dashboard/components/dropdown/stories/index.js (hunk 1 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/components/dropdown/stories/index.js]: sure!

File: assets/src/dashboard/components/dropdown/stories/index.js (hunk 2 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/components/dropdown/stories/index.js]: (updated)

**Review by carloskelly13 — APPROVED** [review]

**Review by mariano-formidable — CHANGES_REQUESTED** [review]
It's looking great, I just noticed three issues I'd like your thoughts on:

1.)  It seems to me that "All Categories" is a special button.  I would expect that when it's selected, all the other pills should unselect.  And if you select another pill whilst "All Categories" is selected, then "All Categories" should unselect itself.  Does that make sense?  I believe there should never exist a state where "All Categories" and another pill are selected at the same time:
image

2.)  If you initialize the panel with an item with `selected: true`, then that option can never be unselected.  In the image below, I've initialized `beauty` with `selected: true`:
pill_selectable

3.)  Feel free to push back on this idea, but I'm not sure if `value` should be used as the determinate of whether a pill is selectable or not (right now it seems that setting `value: false` will make the pill unclickable).  Is there no case where `false` might be a valid value?  Also I think it would be clearer to those using the panel that want to disable certain options to use something like `disabled: true` rather than setting the `value` to a special value... what are you thoughts?  Have you seen this pattern before (i.e. `value: false` means it's disabled)?  If it's the en vogue thing to do for checkboxes then by all means keep it and let me know, I'm all for learning to latest and greatest 😎 

**NOTE:** Feel free to address these in a follow up PR if this PR was focused mostly on UI and not the nitty gritty functionality

File: assets/src/dashboard/components/dropdown/stories/index.js (hunk 2 above)
Comment (mariano-formidable) [reply assets/src/dashboard/components/dropdown/stories/index.js]: Yeah, I'm not sure we need to translate storybook text... I don't see "UNCLICKABLE" being translated any time soon :D

**Comment by BrittanyIRL** [comment]
All great questions! I addressed 2 + 3 and feel like 1 can stand alone. I'm curious what direction you think we should head in for it.

> 1.) It seems to me that "All Categories" is a special button. I would expect that when it's selected, all the other pills should unselect. And if you select another pill whilst "All Categories" is selected, then "All Categories" should unselect itself. Does that make sense? I believe there should never exist a state where "All Categories" and another pill are selected at the same time.

- Yes. I was really focused here on _just_ the ui for the popover panel itself. I used 'categories' as an example so that it was easier to compare in figma. The way I see it we have two options here to handle the 'all' vs. everything else. 
1. We handle that logic outside of the panel, so like only allow 'all' to be selected if no other options are selected. This would keep the panel outside of any logic set for it. 
2. We set up a special prop for the 'all' pill and maintain this logic within the panel itself. Letting the 'all' be an optional setting within the panel for cases where it's moot. 


> 2.) If you initialize the panel with an item with `selected: true`, then that option can never be unselected. In the image below, I've initialized `beauty` with `selected: true`:

-  So, this was a couple things it turns out. All of which are now very good to know about storybook! 1) the popover panel demo in storybook wasn't hooked up to state like the dropdown was. That's fixed - as are the pill demos themselves so that experiences are consistent. The other fun thing here that I didn't realize at fist is that the knobs on storybook compete with the new state for priority. FUN. 
Anyway, this issue should be resolved. Can you look at it again? 


> 3.) Feel free to push back on this idea, but I'm not sure if `value` should be used as the determinate of whether a pill is selectable or not (right now it seems that setting `value: false` will make the pill unclickable). Is there no case where `false` might be a valid value?
-  I'm down to make a disabled prop :D using value as false is just a habit from setting dropdowns up. I'll make this change, i think it is easier to read. no use to double up on prop usage like this unless there's a specific reason.  This should be updated now!

**Review by mariano-formidable — APPROVED** [review]
Okay, I see what's going on now.  The entire panel and popover have no internal react state, the pills could be toggled on and off because they're checkboxes and the browser was taking care of maintaining the checkbox state.  In order to fix the bug I found, you had to introduce react state management at the storybook level so we can toggle things off and on and override browser state...

And with state being controlled outside of the panel, then yes, the "All Categories" logic should be handled by whatever parent component implements the panel.  Okaaay, it's clicking with me now!

Thanks for getting things to work (and added the `disabled` prop), but it definitely wasn't super necessary to add extra logic to storybook to get a thing working, so sorry for missing that and adding extra work to you 😱 

The UI looks great and it'll be easy to use on our site as needed!

**Comment by BrittanyIRL** [comment]
Summarized so well, yes! And no worries, i feel like the continuity is good just as future sanity checks if anything breaks. I like keeping storybook current and this was necessary for that!

**Pull request merged**